src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceDeserializer.cs

```diff
@@ -245,6 +263,76 @@ public virtual object CreateResourceInstance(IEdmStructuredTypeReference structu
             }
         }
 
+        private object DeSerializeDeletedEntity(ODataDeletedResource deletedResource, IEdmStructuredTypeReference structuredType, ODataDeserializerContext readContext, Type clrType)
+        {
+            Contract.Assert(deletedResource != null, "ODataDeletedResource should not be null");
+
+            IEdmModel model = readContext.Model;
+
+            Contract.Assert(model != null);
+
+            string typeName = deletedResource.TypeName;
+
+            IEdmEntityType actualType = model.FindType(typeName) as IEdmEntityType;
+            if (actualType == null)
+            {
+                throw new ODataException(Error.Format(SRResources.ResourceTypeNotInModel, typeName));
+            }
+
+            if (actualType.IsAbstract)
+            {
+                string message = Error.Format(SRResources.CannotInstantiateAbstractResourceType, typeName);
+                throw new ODataException(message);
+            }
+
+            PropertyInfo instanceAnnotationProperty = EdmLibHelpers.GetInstanceAnnotationsContainer(
+                        structuredType.StructuredDefinition(), model);
+
+            dynamic deletedEntity;
+
+            deletedEntity = SetDeletedEntity(deletedResource, readContext, clrType, actualType, instanceAnnotationProperty);
+
+            DeserializationHelpers.ApplyInstanceAnnotations(deletedEntity, actualType.ToEdmTypeReference(false).AsStructured(), deletedResource, DeserializerProvider, readContext);
+
+            return deletedEntity;
+        }
+
+        private dynamic SetDeletedEntity(ODataDeletedResource deletedResource, ODataDeserializerContext readContext, Type clrType, IEdmEntityType actualType, PropertyInfo instanceAnnotationProperty)
+        {
+            dynamic deletedEntity;
+            if (readContext.IsUntyped)
+            {
+                deletedEntity = new EdmDeltaDeletedEntityObject(actualType);
+            }
+            else
+            {
+                Type type = typeof(DeltaDeletedEntityObject<>).MakeGenericType(clrType);
+
+                if (instanceAnnotationProperty == null)
+                {
+                    deletedEntity = Activator.CreateInstance(type, clrType);
+                }
+                else
+                {
+                    deletedEntity = Activator.CreateInstance(type, clrType, instanceAnnotationProperty);
+                }
+            }
+
+            deletedEntity.Id = deletedResource.Id == null ? "" : deletedResource.Id.ToString();
+            deletedEntity.Reason = deletedResource.Reason.Value;
+
+            SetProperties(deletedResource, deletedEntity);
+            return deletedEntity;
+        }
+
+        private static void SetProperties(ODataDeletedResource deletedResource, dynamic deletedEntity)
+        {
+            foreach (ODataProperty property in deletedResource.Properties)
+            {
+                deletedEntity.TrySetPropertyValue(property.Name, property.Value);
+            }
```

Comment:
Could DeletedResource have normal properties?  @mikepizzo 

From OData spec: The deleted-entity object MAY include additional properties of the entity, as well as annotations, and MAY include related entities, related deleted entities, or a delta or full representation of a related collection of entities, to represent related entities that have been modified or deleted.

I think that's why ODataDeletedResource is derived from ODataResourceBase.

So, SetProperties(...) method doesn't reflect this.  It only calls "TrySetPropertyValue", it only works for fewer primitive value, how about others? "related entities, related deleted entities, ...."?